Mode aware consistency check #13467 clean

dcarpentiero opened the pull request:
Closes https://github.com/JabRef/jabref/issues/13467

This draft PR refactors BibliographyConsistencyCheck to take a full BibDatabaseContext instead of just a list of entries.
The goal is to prepare the logic for BibTeX/BibLaTeX-aware consistency checking, as discussed in the issue.
So far, logic separation for mode (BibDatabaseMode) and extraction of EntryType has been implemented.

Steps to test
This is a work-in-progress. No final test procedure yet.
Initial behavior should be unchanged (functional refactoring).
Can be tested with a .bib file containing @online entries with and without date.

Mandatory checks
 I own the copyright ...
[/] Change in CHANGELOG.md ...
 Tests created ...
[/] Manually tested ...
[/] Screenshots added ...
[/] Checked developer's documentation ...
[/] Checked documentation ...

jabref-machine commented:
You have removed the "Mandatory Checks" section from your pull request description. Please adhere to our [pull request template](https://github.com/JabRef/jabref/blob/main/.github/PULL_REQUEST_TEMPLATE.md?plain=1#L10).


<!-- ghprcomment
PR Tests
Mandatory Checks present
-->

jabref-machine commented:
Note that your PR will not be reviewed/accepted until you have gone through the mandatory checks in the description and marked each of them them exactly in the format of `[x]` (done), `[ ]` (not done yet) or `[/]` (not applicable).


<!-- ghprcomment
PR Tests
PR checklist OK
-->

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Source Code Tests
Checkstyle
-->

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Source Code Tests
Checkstyle
-->

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jablib
-->

dcarpentiero commented:
Hello all, I'd like to get some help regarding a failing test in `jablib` which seems unrelated to my modifications. Since this seems to fall outside the scope of the issue I'm addressing, how would you like me to proceed? 
Thanks for the replies!

subhramit commented:
> @subhramit hi that's me. I deleted the old branch with the old rebase problem and i've started clean.

It's fine this time - but for future, we generally don't recommend closing and opening - one can update the existing branch so that the PR is also updated.

subhramit commented:
> Hello all, I'd like to get some help regarding a failing test in `jablib` which seems unrelated to my modifications. Since this seems to fall outside the scope of the issue I'm addressing, how would you like me to proceed?
> Thanks for the replies!

Open the failing check by clicking on it, compare the expected and actual values in the log.
Then trace back how your changes caused this.

As far as I can see:
```
org.jabref.logic.quality.consistency.BibliographyConsistencyCheckResultCsvWriterTest

 Test checkComplexLibrary(Path) FAILED

  org.opentest4j.AssertionFailedError: expected: <entry type,citation key,Location,Pages,Publisher
  Article,first,-,o,-
  Article,second,-,-,?
  InProceedings,fourth,-,-,o
  InProceedings,third,?,o,-
  > but was: <entry type,citation key,Pages,Publisher
  Article,first,o,-
  Article,second,-,?
  >
```
Some things seem to be missing from the actual.
```
org.jabref.logic.quality.consistency.BibliographyConsistencyCheckTest

  Test unsetFieldsReported() FAILED

  org.opentest4j.AssertionFailedError: expected: <[@online{withDate,
    date = {date},
    urldate = {urldate},
    _jabref_shared = {sharedId: -1, version: 1}
  }, @online{withoutDate,
    urldate = {urldate},
    _jabref_shared = {sharedId: -1, version: 1}
  }]> but was: <[@online{withDate,
    date = {date},
    urldate = {urldate},
    _jabref_shared = {sharedId: -1, version: 1}
  }]>
```

dcarpentiero commented:
@subhramit I don't know I think I found it like this if I remember correctly. BTW i removed the "..."
tomorrow I will resolve the test issues, thanks for helping out

dcarpentiero commented:
> > @subhramit I don't know I think I found it like this if I remember correctly. BTW i removed the "..."
> 
> Check the other PRs to see the unmodified version

Ohh thanks I see... I corrected it!

calixtus commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -105,10 +110,41 @@ public Result check(List<BibEntry> entries, BiConsumer<Integer, Integer> entries
+            entryTypeToEntriesMap
+                    .computeIfAbsent(entryType, _ -> new HashSet<>())
+                    .add(entry);
+                } else if (mode == BibDatabaseMode.BIBTEX && bibtexSet.contains(entry.getType())) {
```
indent seems off

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jablib
-->

dcarpentiero commented:
@subhramit good morning, what else is left to do to merge to main and close pull request? all tests have passed.

koppor commented on jabgui/src/main/java/org/jabref/gui/consistency/ConsistencyCheckAction.java:
```diff
@@ -55,10 +53,11 @@ public BibliographyConsistencyCheck.Result call() throws Exception {
                 }
-                List<BibEntry> entries = databaseContext.get().getEntries();
+
+                BibDatabaseContext bibContext = databaseContext.orElseThrow(() -> new IllegalStateException("database not present"));
```
Never throw `IllegalStateException`, because this tears down JabRef. Just use `.get()` here.

koppor commented on jabgui/src/main/java/org/jabref/gui/consistency/ConsistencyCheckAction.java:
```diff
@@ -55,10 +53,11 @@ public BibliographyConsistencyCheck.Result call() throws Exception {
                 if (databaseContext.isEmpty()) {
                     throw new IllegalStateException((Localization.lang("No library present")));
```
Replace by throwing a `JabRefException`.

dcarpentiero replied:
done.

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -105,10 +110,41 @@ public Result check(List<BibEntry> entries, BiConsumer<Integer, Integer> entries
+        BibDatabaseMode mode = bibContext.getMode();
+        List<BibEntry> entries = bibContext.getEntries();
+
+        Set<EntryType> biblatexSet = Set.of();
+        Set<EntryType> bibtexSet = Set.of();
```
Can't this be unified to `typeSet`? - so that one needs only one set?

I think, there is no initializiation required.

dcarpentiero replied:
done.

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -105,10 +110,41 @@ public Result check(List<BibEntry> entries, BiConsumer<Integer, Integer> entries
+        Set<EntryType> biblatexSet = Set.of();
+        Set<EntryType> bibtexSet = Set.of();
+
+        if (mode == BibDatabaseMode.BIBLATEX) {
```
mode is an `enum`, isn't it? A switch should work?!

dcarpentiero replied:
refactor done.

dcarpentiero commented on jablib/src/test/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheckTest.java:
```diff
@@ -143,12 +175,18 @@ void unsetFieldsReported() {
+        BibDatabase bibDatabase = new BibDatabase(bibEntriesList);
+        bibDatabase.insertEntries(bibEntriesList);
+
+        BibDatabaseContext bibContext = new BibDatabaseContext(bibDatabase);
```
that's the step i've missed. thanks again

jabref-machine commented:
JUnit tests of `jabkit` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jabkit
-->

dcarpentiero commented:
@koppor I was running the CLI tests and noticed something about the checkConsistencyPorcelain() test.
It expects the origin.bib file to produce an empty output (""), but right now that file contains entries (like Newton or Einstein), so the output isn't empty anymore, it's a small markdown table.
This makes the test break if the .bib file changes, even a little. Maybe it would be better if the test checked for the presence of a markdown table or just verified that the output is in the expected format? how should i proceed?

koppor commented:
> @koppor I was running the CLI tests and noticed something about the checkConsistencyPorcelain() test. It expects the origin.bib file to produce an empty output (""), but right now that file contains entries (like Newton or Einstein), so the output isn't empty anymore, it's a small markdown table. This makes the test break if the .bib file changes, even a little. Maybe it would be better if the test checked for the presence of a markdown table or just verified that the output is in the expected format? how should i proceed?

I think, something is wrong in the logic. The output of the consistency check is

| entry type | citation key | Author | Citationkey | Publisher | Title | Year |
  | ---------- | ------------ | ------ | ----------- | --------- | ----- | ---- |
  | Book       | Darwin1888   | ?      | ?           | ?         | ?     | ?    |
  | Book       | Einstein1920 | ?      | ?           | ?         | ?     | ?    |
  | Book       | Newton1999   | ?      | ?           | ?         | ?     | ?    |

The `?` means that the field is unkown. However, `@Book` should be a required type.

Moreover, "citatoin key" appears twice.

--> Something is wrong in the logic

jabref-machine commented:
JUnit tests of `jabkit` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jabkit
-->

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jablib
-->

koppor commented on jablib/src/test/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheckTest.java:
```diff
@@ -143,8 +176,15 @@ void unsetFieldsReported() {
+        bibDatabase.insertEntries(bibEntriesList);
+
+        BibDatabaseContext bibContext = new BibDatabaseContext(bibDatabase);
+        bibContext.setMode(BibDatabaseMode.BIBLATEX);
```
There should be a second test with `BIBTEX` mode.

Reason `ONLINE` is only defined for BibLaTeX: org.jabref.model.entry.types.BiblatexEntryTypeDefinitions#ONLINE, thus, this behaves different in bibtex and biblatex.

BibLaTeX:

| Field | Required |
| -- | -- |
| DATE | yes |
| URLDATE | no (optional) |

Thus, `withoutDate` misses the required field `date` and the output should be that `withoutDate` is reported.

`withDate`  is reported, because requried fields author/editor, title, and URL are missing.

dcarpentiero replied:
i've added it now

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jablib
-->

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -62,64 +82,115 @@ public record EntryTypeResult(Collection<Field> fields, SequencedCollection<BibE
+        List<BibEntry> entries = bibContext.getEntries();
+
+        Set<EntryType> typeSet = switch (mode) {
+            case BIBLATEX -> BiblatexEntryTypeDefinitions.ALL.stream()
+                    .map(BibEntryType::getType)
+                    .collect(Collectors.toSet());
```
`BiblatexEntryTypeDefinitions.ALL` is a constant --> extract this into a static field

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -62,64 +82,115 @@ public record EntryTypeResult(Collection<Field> fields, SequencedCollection<BibE
+            case BIBLATEX -> BiblatexEntryTypeDefinitions.ALL.stream()
+                    .map(BibEntryType::getType)
+                    .collect(Collectors.toSet());
+            case BIBTEX -> BibtexEntryTypeDefinitions.ALL.stream()
```
`BibtexEntryTypeDefinitions.ALL` is a constant --> extract this into a static field

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -14,16 +15,22 @@
 public class BibliographyConsistencyCheck {
 
     private static final Set<Field> EXPLICITLY_EXCLUDED_FIELDS = Set.of(
+                            StandardField.KEY,
```
I am not sure about this one, but I thinkj, it is OK to leave this in. 

Also include all from "JabRef-specific fields" (org.jabref.model.entry.field.StandardField#CITATIONCOUNT etc)

Also includde `org.jabref.model.entry.field.InternalField#KEY_FIELD`.

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -62,64 +82,115 @@ public record EntryTypeResult(Collection<Field> fields, SequencedCollection<BibE
-            entryTypeToEntriesMap
-                    .computeIfAbsent(entryType, _ -> new HashSet<>())
-                    .add(entry);
+                if (!entryTypeToFieldsInAllEntriesMap.containsKey(entryType)) {
```
Swap branches --> negations in if conditions should be avoided

Add a comment why this is the flow - I cannot get it by reading.

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -62,64 +82,115 @@ public record EntryTypeResult(Collection<Field> fields, SequencedCollection<BibE
+            }
+
+            List<BibEntry> sortedEntries;
+            if (mode == BibDatabaseMode.BIBLATEX) {
```
Why this? BibTeX also has required and optional fields. -- I think, the if is not required.

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -62,64 +82,115 @@ public record EntryTypeResult(Collection<Field> fields, SequencedCollection<BibE
 
-            resultMap.put(entryType, new EntryTypeResult(uniqueFields, differingEntries));
+            Set<BibEntry> entries = entryTypeToEntriesMap.get(entryType);
+            if (entries == null || entries.size() <= 1 || differingFields.isEmpty()) {
```
I think, also one differenting entry should be output? --> maybe craft a test case hitting exactly this example.

koppor replied:
I think, the check is OK. However, I think, this case never happens. Therefore, I added an `assert` stgatement.

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -62,64 +82,115 @@ public record EntryTypeResult(Collection<Field> fields, SequencedCollection<BibE
-                    .filter(entry -> !filterExcludedFields(entry.getFields()).equals(commonFields))
-                    .sorted(comparatorStack)
-                    .toList();
+            if (mode == BibDatabaseMode.BIBLATEX) {
```
No. BibTeX also has required fields.. Should make no difference. in handling.

koppor reviewed (changes requested):
Maybe, we need more tests

bibtex / biblatex

dcarpentiero commented:
Hi @koppor,

After implementing your suggested refactorings (constants extraction, field exclusions, unified BibTeX/BibLaTeX logic), the tests are failing due to inconsistent expectations in the entry filtering behavior.

**Current Issue:**
The `filterEntriesWithFieldDifferences` method needs clarification on which entries to include when field inconsistencies are detected within an entry type.

**Specific Test Conflicts:**
- `unsetFieldsReported`: expects both `withDate` and `withoutDate` entries ✅
- `checkDifferentOutputSymbols`: expects only `first` entry (has differing fields) ❌
- `checkComplexLibrary`: expects only entries with actual field differences ❌

**Question:** When there are differing fields in an entry type, should we include:
1. **Only entries that have at least one differing field present** (current logic)
2. **All entries of that type** (since they're all inconsistent by definition)
3. **Different logic based on field type** (required vs optional)?

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jablib
-->

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -45,6 +63,17 @@ private static Set<Field> filterExcludedFields(Collection<Field> fields) {
+    private static List<BibEntry> filterEntriesWithFieldDifferences(Set<BibEntry> entries, Set<Field> differingFields, Set<Field> fieldsInAllEntries) {
+        // For required fields (when the BibliographyConsistencyCheck adds them to differingFields),
+        // we need to include ALL entries that participate in the inconsistency.
+        // For regular fields, we only include entries that actually have the differing fields.
```
There is no "regular" in JabRefs language. -- If you want to know more, read the biblatex manual https://ctan.org/pkg/biblatex

```suggestion
        // For optional and custom fields, we only include entries that actually have the differing fields.
```

You can also use "non-required, but present"

koppor replied:
Line 80 says

>     * This class <em>does not</em> check whether all required fields are present or if the fields are valid for the entry type.

I think, this can be kept? - thus, no difference in handling?

koppor replied:
Move the description as method documentaiton.

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -45,6 +63,17 @@ private static Set<Field> filterExcludedFields(Collection<Field> fields) {
+        // If this method is called, it means there are field differences that need to be reported
+        // Since the logic above already determines which fields are differing and adds required fields
+        // when they're inconsistent, we should include all entries to show the complete picture
+        return new ArrayList<>(entries);
```
This is a TODO - otherwise, the method does not do what the description says.

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jablib
-->

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -33,7 +49,9 @@ public class BibliographyConsistencyCheck {
-                            StandardField.XREF
+                            StandardField.XREF,
+                            StandardField.CITATIONCOUNT, // JabRef-specific
+                            InternalField.KEY_FIELD      // Internal field
```
```suggestion
                            InternalField.KEY_FIELD      // Citation key
```

koppor commented:
I think, my last comment got lost. I try to rewrite.

- GUI: Users should see "nice" strings
- BibTeX (also BibTeX source tab): good unique serialization

Meaning: `s1\,s2` should be displayed as "s1,s2" in one of these clickable things in the UI, but written as `s1\,s2` to the bib file

Assuming that `asString` is for seralization, I began to add following changes. I had no time to continue yet. If you have: good. But be warned, it might lead to a hole refacotring of the keyword and keyword list thing. One really needs to be aware when something is read from the bibtex and written into the bibtex.

Maybe, there have to be different methods inside reader and writer of bibtex files - and the UI handling can be "normal"? - This especially refs https://github.com/JabRef/jabref/issues/11390

```diff
diff --git a/jablib/src/main/java/org/jabref/model/entry/Keyword.java b/jablib/src/main/java/org/jabref/model/entry/Keyword.java
index d3df45becc..3f416fa045 100644
--- a/jablib/src/main/java/org/jabref/model/entry/Keyword.java
+++ b/jablib/src/main/java/org/jabref/model/entry/Keyword.java
@@ -2,6 +2,8 @@ package org.jabref.model.entry;

 import java.util.Objects;
 import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;

@@ -13,7 +15,9 @@ import org.jabref.model.ChainNode;
  */
 public class Keyword extends ChainNode<Keyword> implements Comparable<Keyword> {

+    // Note: {@link org.jabref.model.entry.KeywordList#parse(java.lang.String, java.lang.Character, java.lang.Character) offers configuration, which is not available here
     public static Character DEFAULT_HIERARCHICAL_DELIMITER = '>';
+
     private final String keyword;

     public Keyword(String keyword) {
@@ -79,7 +83,7 @@ public class Keyword extends ChainNode<Keyword> implements Comparable<Keyword> {
      * E.g., calling {@link #getSubchainAsString(Character)} on the node "B" in "A > B > C" returns "B > C".
      */
     private String getSubchainAsString(Character hierarchicalDelimiter) {
-        return keyword +
+        return getEscaped(hierarchicalDelimiter) +
                 getChild().map(child -> " " + hierarchicalDelimiter + " " + child.getSubchainAsString(hierarchicalDelimiter))
                           .orElse("");
     }
@@ -99,7 +103,14 @@ public class Keyword extends ChainNode<Keyword> implements Comparable<Keyword> {
         return getParent()
                 .map(parent -> parent.getPathFromRootAsString(hierarchicalDelimiter) + " " + hierarchicalDelimiter + " ")
                 .orElse("")
-                + keyword;
+                + getEscaped(hierarchicalDelimiter);
+    }
+
+    public String getEscaped(Character hierarchicalDelimiter) {
+        String escapedDelimiter = Pattern.quote(String.valueOf(hierarchicalDelimiter));
+        Pattern pattern = Pattern.compile("(?<!\\\\)" + escapedDelimiter);
+        Matcher matcher = pattern.matcher(keyword);
+        return matcher.replaceAll("\\" + hierarchicalDelimiter);
     }

     /**
diff --git a/jablib/src/main/java/org/jabref/model/entry/KeywordList.java b/jablib/src/main/java/org/jabref/model/entry/KeywordList.java
index febc1e0134..dce6f1b7dc 100644
--- a/jablib/src/main/java/org/jabref/model/entry/KeywordList.java
+++ b/jablib/src/main/java/org/jabref/model/entry/KeywordList.java
@@ -97,7 +97,7 @@ public class KeywordList implements Iterable<Keyword> {
     }

     public static String serialize(List<Keyword> keywords, Character delimiter) {
-        return keywords.stream().map(Keyword::get).collect(Collectors.joining(delimiter.toString()));
+        return keywords.stream().map(keyword -> keyword.getEscaped(delimiter)).collect(Collectors.joining(delimiter.toString()));
     }

```

dcarpentiero commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -73,7 +73,6 @@ private static Set<Field> filterExcludedFields(Collection<Field> fields) {
      * @param fieldsInAllEntries fields that are present in all entries
      * @return a list of entries that present at least one of the differing fields
      */
-    @VisibleForTesting
```
if i apply the annotation "VisibleForTesting" this error comes up: "> Task :jabgui:test org.jabref.gui.JabGuiArchitectureTests

Test shouldUseJSpecifyAnnotations FAILED

java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'no classes should depend on classes that reside in a package 'org.jetbrains.annotations..', because JSpecify annotations should be used' was violated (1 times): Method <org.jabref.logic.quality.consistency.BibliographyConsistencyCheck.filterEntriesWithFieldDifferences(java.util.Set, java.util.Set, java.util.Set)> is annotated with <org.jetbrains.annotations.VisibleForTesting> in (BibliographyConsistencyCheck.java:0"

koppor reviewed (changes requested):
Please adress the previosu comments. THEN re-request review. Please!!

Here a screenshot of the unadressed comemnt - in case your browser does not render https://github.com/JabRef/jabref/pull/13584/files properly.

<img width="951" height="812" alt="image" src="https://github.com/user-attachments/assets/a832b5d3-b445-439b-b138-4368dfc001f1" />

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -14,27 +16,50 @@
+            StandardField.TYPE,
+            StandardField.XREF,
+            StandardField.OWNER,
+            StandardField.CITATIONCOUNT, // JabRef-specific
```
I think, there is a learning gap at how comments are used - and what "consistency" means in source code.

<img width="670" height="258" alt="image" src="https://github.com/user-attachments/assets/b97ae467-a3b7-4eb8-ac21-c2fe486529f9" />

I fixed in 5b41900b66 - maybe, you see using the commit, what I mean.

koppor commented on jablib/src/test/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheckTest.java:
```diff
@@ -143,12 +179,87 @@ void unsetFieldsReported() {
+        );
+
+        BibEntry entry1 = new BibEntry(StandardEntryType.Article, "id1")
+                .withField(StandardField.AUTHOR, "Autore Uno")
```
I like more diversity in the code. This is inderstandable for most persons, thus, it is OK to leave it in.

In general, English language should be used also in examples -. if the exmaples do not touch some non-English language handling.

koppor commented on jablib/src/test/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheckResultTxtWriterTest.java:
```diff
@@ -178,21 +211,21 @@ void checkComplexLibrary(@TempDir Path tempDir) throws IOException {
                 .withField(StandardField.AUTHOR, "Author One")
                 .withField(StandardField.YEAR, "2024")
                 .withField(StandardField.PUBLISHER, "publisher");
-        // Entry added to check for alphabetical ordering of citation keys
         BibEntry fifth = new BibEntry(StandardEntryType.InProceedings, "fifth")
-                .withField(StandardField.AUTHOR, "Author One")
-                .withField(StandardField.LOCATION, "location")
-                .withField(StandardField.YEAR, "2024");
-        BibEntry sixth = new BibEntry(StandardEntryType.InProceedings, "sixth")
                 .withField(StandardField.AUTHOR, "Author One")
```
I think, you wanted to add **another** test case.

--> (in General) DO NOT modify test input data when tests do not pass.

Action point:

Duplciate the test.

The first version should have the input as the original test. With that one can see at the diff how the output changed FOR THE SAME input.

dcarpentiero replied:
ok working on that 👍

koppor reviewed (changes requested):
Two comments on tests. One on test modification and one one coverage. Hope, you can adress it.

---

Minor comment: Do you use IntelliJ? Did you notice the "highlighing" in gray?

<img width="723" height="223" alt="image" src="https://github.com/user-attachments/assets/0ae08ed0-7dcb-44eb-b216-f948d7ab8c5b" />

Applied in a066646490

koppor commented:
> Two comments on tests. One on test modification and one one coverage. Hope, you can adress it.

You do know how to trigger coverage?

<img width="1017" height="581" alt="image" src="https://github.com/user-attachments/assets/86e1d707-08f4-4070-85a7-bf736d6bd77c" />

dcarpentiero commented:
I've run all tests with coverage within org.jabref.logic.quality.consistency package, as per requested. they passed, these are the results.

<img width="918" height="444" alt="Screenshot 2025-08-04 alle 09 12 25" src="https://github.com/user-attachments/assets/2adce85c-7def-403e-b902-8f085d51cf4b" />

koppor commented on jablib/src/test/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheckResultTxtWriterTest.java:
```diff
@@ -178,21 +211,21 @@ void checkComplexLibrary(@TempDir Path tempDir) throws IOException {
                 .withField(StandardField.AUTHOR, "Author One")
                 .withField(StandardField.YEAR, "2024")
                 .withField(StandardField.PUBLISHER, "publisher");
-        // Entry added to check for alphabetical ordering of citation keys
```
This comment is still shown as deleted

dcarpentiero replied:
Added back.

koppor commented on jablib/src/test/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheckResultTxtWriterTest.java:
```diff
@@ -178,21 +211,21 @@ void checkComplexLibrary(@TempDir Path tempDir) throws IOException {
                 .withField(StandardField.PUBLISHER, "publisher");
-        // Entry added to check for alphabetical ordering of citation keys
         BibEntry fifth = new BibEntry(StandardEntryType.InProceedings, "fifth")
-                .withField(StandardField.AUTHOR, "Author One")
-                .withField(StandardField.LOCATION, "location")
-                .withField(StandardField.YEAR, "2024");
-        BibEntry sixth = new BibEntry(StandardEntryType.InProceedings, "sixth")
```
Still deletions in.

Not sure if you can easily fix this.

subhramit replied:
Why were these lines (183-186) deleted in the first place?

calixtus commented:
Hi, whats the status here? Are you planning to continue your work on this PR?
This might become an important and good addition.

dcarpentiero commented:
Yes I’m planning to continue, I’m on holiday right now and can work again
on this on Monday. Could you wait still three more days?

Il giorno ven 22 ago 2025 alle 10:57 Carl Christian Snethlage <
***@***.***> ha scritto:

> *calixtus* left a comment (JabRef/jabref#13584)
> <https://github.com/JabRef/jabref/pull/13584#issuecomment-3213616371>
>
> Hi, whats the status here? Are you planning to continue your work on this
> PR?
> This might become an important and good addition.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/13584#issuecomment-3213616371>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBQ63CKX5OVZ3575WLW6S633O3LOFAVCNFSM6AAAAACCITUCOCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMJTGYYTMMZXGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

dcarpentiero commented:
Yes I’m planning to continue, I’m on holiday right now and can work again on this on Monday. Could you wait still three more days?

calixtus commented:
Sure, no problem, i was just doing chores and was checking the recent PRs if they were stalling.

dcarpentiero commented on jablib/src/test/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheckTest.java:
```diff
@@ -262,4 +262,44 @@ void checkFieldEntriesWithFieldDifferences() {
+        BibEntry second = new BibEntry(StandardEntryType.Article, "second")
+                .withField(StandardField.AUTHOR, "Author One")
+                .withField(StandardField.PUBLISHER, "publisher");
+
```
removed empty line

koppor replied:
> removed empty line

Thank you! And sorry for the back-and-forth. I am assuming you would like to learn something about software engineering while contributing 😅. I should have some time this week to support so that we can finally merge!

jabref-machine commented:
JUnit tests of `jabsrv` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jabsrv
-->

jabref-machine commented:
JUnit tests of `jabkit` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jabkit
-->

dcarpentiero commented:
@subhramit hello I might need your help with this one. On the latest commit, there is an error which says that there's not a Java installation on the machine matching `languageVersion=24, vendor=Amazon Corretto, implementation=vendor-specific, nativeImageCapable=false`, what can I do?

subhramit commented:
> @subhramit hello I might need your help with this one. On the latest commit, there is an error which says that there's not a Java installation on the machine matching `languageVersion=24, vendor=Amazon Corretto, implementation=vendor-specific, nativeImageCapable=false`, what can I do?
> 
> Edit: These build errors appear on GitHub, not local

Can you tell me what exactly you were trying to achieve in the [latest commit](https://github.com/JabRef/jabref/pull/13584/commits/ff523826c7e2b5f9001b312405451e462856ca8b)?

dcarpentiero commented:
> > @subhramit hello I might need your help with this one. On the latest commit, there is an error which says that there's not a Java installation on the machine matching `languageVersion=24, vendor=Amazon Corretto, implementation=vendor-specific, nativeImageCapable=false`, what can I do?
> > Edit: These build errors appear on GitHub, not local
> 
> Can you tell me what exactly you were trying to achieve in the [latest commit](https://github.com/JabRef/jabref/pull/13584/commits/ff523826c7e2b5f9001b312405451e462856ca8b)?

I was trying to install Amazon Corretto 24 on the CI runner, it says it was missing so it was the first thing it came to my mind. Given the fact it didn't work, I removed the file though.

dcarpentiero commented:
> > I was trying to install Amazon Corretto 24 on the CI runner, it says it was missing so it was the first thing it came to my mind. Given the fact it didn't work, I removed the file though.
> 
> Oh, did we request working on the CI? As far as I can see, the current workflows of `main` run: https://github.com/JabRef/jabref/actions?query=branch%3Amain
> 
> But I see that your PR does not work:
> 
> ```
>  > Error while evaluating property 'javaCompiler' of task ':jablib:compileJava'.
>    > Failed to calculate the value of task ':jablib:compileJava' property 'javaCompiler'.
>       > Cannot find a Java installation on your machine (Linux 6.11.0-1018-azure amd64) matching: {languageVersion=24, vendor=Amazon Corretto, implementation=vendor-specific, nativeImageCapable=false}. No matching toolchain could be found in the configured toolchain download repositories.
> ```
> 
> I asked GitHub to rerun the jobs -- the error is gone. Its a temporary issue.

Ok sorry then, my mistake. I completed the checks, may we merge to main?

Siedlerchr commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,29 @@
+          java -version
+          ./gradlew -v
+
+      - name: Run checks
```
Any reason for this workflow?

dcarpentiero replied:
As other have told me, and as I have acknowledged my mistake, no not really. Can the changes be reverted?

koppor replied:
I see three options:

1. Use GitHub UI for this
2. You use GitHub UI for this (and then pull)
3. You delete the file and commit

Which one works best for you?

---

Screenshot for option 1 and 2:

<img width="1214" height="862" alt="image" src="https://github.com/user-attachments/assets/77736396-b116-4424-bf69-dbc1b39d7bc6" />

Powered by https://github.com/refined-github/refined-github

calixtus commented on jabgui/src/main/java/org/jabref/gui/consistency/ConsistencyCheckAction.java:
```diff
@@ -53,12 +52,13 @@ public BibliographyConsistencyCheck.Result call() throws Exception {
                 Optional<BibDatabaseContext> databaseContext = stateManager.getActiveDatabase();
                 if (databaseContext.isEmpty()) {
-                    throw new IllegalStateException((Localization.lang("No library present")));
+                    throw new JabRefException("No library present.");
```
This Exception has to be removed completly. Not having a library opened is not an exceptional state but a legal state of jabref. If the user clicks around or for any other reason the consistency check is run in the gui, this should NOT throw an exception. Exceptions are for exceptional states only.
Please remove, make a simple dialog message maybe and write sthg to debug log.

koppor replied:
`dialogService.notify` should be enough --> reason: We don't want to break the users flow and try to use as less dialogs as possible.

(We are aware that these toast messages should be collected somehow, but this is future work)

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jablib
-->

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jablib
-->

koppor commented:
We use this PR to fix https://github.com/JabRef/jabref/pull/13765

calixtus reviewed (approved):
Looks fine to me. Lets get this merged.

dcarpentiero commented:
Thanks for the opportunity to learn.

dcarpentiero commented:
Thanks to the opportunity to learn. I'd like to ask, could I appear in the contributors list? Thank you.

subhramit commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -62,64 +110,94 @@ public record EntryTypeResult(Collection<Field> fields, SequencedCollection<BibE
+        }
 
-        Map<EntryType, EntryTypeResult> resultMap = new HashMap<>();
+        // Use LinkedHashMap to preserve the order of Bib*EntryTypeDefinitions.ALL
```
Okay, it is.

koppor commented on jablib/src/main/java/org/jabref/logic/quality/consistency/BibliographyConsistencyCheck.java:
```diff
@@ -62,64 +110,94 @@ public record EntryTypeResult(Collection<Field> fields, SequencedCollection<BibE
+                }
+
+                entryTypeToEntriesMap
+                        .computeIfAbsent(entryType, _ -> new java.util.LinkedHashSet<>())
```
Oh, yes. Follow-up? ^^
